### PR TITLE
[azure-*-cpp] Shorten source paths

### PR DESCRIPTION
**Pull request by antkmsft**
Windows has a path limit of 260 characters, and certain packages fail to build when depending on azure-messaging-eventhubs.
This change patches the directory names and it does not change any public surface available to the port users.

cc @LarryOsterman

### CODE REVIEW

File: ports/azure-messaging-eventhubs-cpp/portfile.cmake [hunk 1]
```diff
@@ -3,10 +3,12 @@ vcpkg_from_github(
     REPO Azure/azure-sdk-for-cpp
     REF azure-messaging-eventhubs_1.0.0-beta.2
     SHA512 3677edf50218c39e378a968661c0ec0600b8760d3af33f204f3267e104449c0aef42b9520cb63b5e5810b89bf91f06979d009ffdd6d6d04d2c14e1a21f0e111f
+    PATCHES
+        00100-shorten_paths.patch
```
Comment (dg0yt) [inline ports/azure-messaging-eventhubs-cpp/portfile.cmake:7]: For renaming, a CMake command would be more efficient.

File: ports/azure-messaging-eventhubs-cpp/portfile.cmake [hunk 2]
```diff
@@ -3,10 +3,12 @@ vcpkg_from_github(
     REPO Azure/azure-sdk-for-cpp
     REF azure-messaging-eventhubs_1.0.0-beta.2
     SHA512 3677edf50218c39e378a968661c0ec0600b8760d3af33f204f3267e104449c0aef42b9520cb63b5e5810b89bf91f06979d009ffdd6d6d04d2c14e1a21f0e111f
+    PATCHES
+        00100-shorten_paths.patch
 )
 
 vcpkg_cmake_configure(
-    SOURCE_PATH "${SOURCE_PATH}/sdk/eventhubs/azure-messaging-eventhubs/"
+    SOURCE_PATH "${SOURCE_PATH}/sdk/eventhubs/eventhubs/"
```
Comment (dg0yt) [inline ports/azure-messaging-eventhubs-cpp/portfile.cmake:11]: Why not shorten even further? It is just the source dir.

If you don't need anything above `SOURCE_PATH`, you might use more radical renaming before calling `vcpkg_configure_make`.

**Comment by Neumann-A** [comment]
I don't think this is something vcpkg needs to patch. It is up to upstream to change the dir layout. Just put in a length warning and/or use a similar approach to qtwebeengine without needing to add more than 200 lines of patch

**Comment by Cheney-W** [comment]
@antkmsft Could you modify your PR according to dg0yt's suggestions?

**Comment by antkmsft** [comment]
@Neumann-A, I am from the upstream. We will shorten the path in our repo. We don't want to create a new release just to shorten the path without visible user benefits. We don't want to wait for the next planned release. Some of our development is blocked by this, as we made several CI pipelines that obtain our packages via vcpkg on multiple platforms, for better verification.

File: ports/azure-messaging-eventhubs-cpp/portfile.cmake (hunk 2 above)
Comment (antkmsft) [reply ports/azure-messaging-eventhubs-cpp/portfile.cmake:11]: We will be shortening the directory tree in azure-sdk-for-cpp repo the exact same way for the next release, so this port update brings that exact directory structure sooner. For the next release, no patch will be needed.
Current shortening is sufficient, if we're taking about the repo structure, and if we're talking about only having more aggressive shortening in vcpkg, it is additional maintenance cost, while, as I said, the proposed shortening is sufficient at the moment.

**Comment by Neumann-A** [comment]
@antkmsft Use cmake variable `VCPKG_INSTALL_OPTIONS` and add `--x-buildtrees-root=<shortpath>` to shorten the default buildtree vcpkg uses. This PR seems totally unnecessary from my point of view.

**Comment by LarryOsterman** [comment]
> I don't think this is something vcpkg needs to patch. It is up to upstream to change the dir layout. Just put in a length warning and/or use a similar approach to qtwebeengine without needing to add more than 200 lines of patch

For context, the filename lengths are causing issues on Windows which block active development of the azure-sdk-for-cpp repo. 

We don't have the option of making a fix in our CI pipeline right now, because the azure-sdk-for-cpp repo releases package updates on a monthly cadence, but we *can* do this in a patch to vcpkg which will resolve the issue without breaking customers.

And then in the October update, we'll rename the source layout to alleviate the problem long-term.

**Comment by LarryOsterman** [comment]
> @antkmsft Use cmake variable `VCPKG_INSTALL_OPTIONS` and add `--x-buildtrees-root=<shortpath>` to shorten the default buildtree vcpkg uses. This PR seems totally unnecessary from my point of view.

How much of the path does that actually shorten? It looks like the vast majority of the path comes from the source tree. Or is the idea to move the buildtrees-root outside the cmake build scope?

**Comment by Neumann-A** [comment]
> How much of the path does that actually shorten?

that depends on where your vcpkg is located. It basically removes `%vcpkg_root%/buildtrees` and replaces it with `<shortpath>`

File: ports/azure-messaging-eventhubs-cpp/portfile.cmake (hunk 1 above)
Comment (antkmsft) [reply ports/azure-messaging-eventhubs-cpp/portfile.cmake]: Done.

**Comment by antkmsft** [comment]
@Cheney-W, I updated the PR to use a cmake directive for renames, and I added azure storage ports as well.

File: ports/azure-messaging-eventhubs-cpp/portfile.cmake [hunk 3]
```diff
@@ -5,8 +5,10 @@ vcpkg_from_github(
     SHA512 3677edf50218c39e378a968661c0ec0600b8760d3af33f204f3267e104449c0aef42b9520cb63b5e5810b89bf91f06979d009ffdd6d6d04d2c14e1a21f0e111f
 )
 
+file(RENAME ${SOURCE_PATH}/sdk/eventhubs/azure-messaging-eventhubs/ ${SOURCE_PATH}/sdk/eventhubs/eventhubs/)
```
Comment (dg0yt) [inline ports/azure-messaging-eventhubs-cpp/portfile.cmake:8]:
```suggestion
if(EXISTS "${SOURCE_PATH}/sdk/eventhubs/azure-messaging-eventhubs")
    file(REMOVE_RECURSE "${SOURCE_PATH}/sdk/eventhubs/eventhubs")
    file(RENAME "${SOURCE_PATH}/sdk/eventhubs/azure-messaging-eventhubs" "${SOURCE_PATH}/sdk/eventhubs/eventhubs")
endif()
```
The `if` guard is for `--editable` installations.

Comment (dg0yt) [reply ports/azure-messaging-eventhubs-cpp/portfile.cmake:8]: I understand that you are in a hurry to unblock *your CI*.
But I suggested more radical shortening (than just 16 chars) in order to help *your users*. Which may also be unaware of `--x-buildtrees-root`.

```suggestion
if(EXISTS "${SOURCE_PATH}/sdk/eventhubs/azure-messaging-eventhubs")
    file(REMOVE_RECURSE "${SOURCE_PATH}.renamed")
    file(RENAME ${SOURCE_PATH}/sdk/eventhubs/azure-messaging-eventhubs" "${SOURCE_PATH}.renamed")
    file(REMOVE_RECURSE "${SOURCE_PATH}")
    file(RENAME ${SOURCE_PATH}.renamed" "${SOURCE_PATH}")
endif()
```
assuming that nothing outside `azure-messaging-eventhubs` is needed.

Comment (antkmsft) [reply ports/azure-messaging-eventhubs-cpp/portfile.cmake:8]: Thanks! I'll add the `if` guard, but we don't want to shorten the number of levels in directories - on one occasion, library's cmake does include a sibling header-only library as `../azure-security-keyvault-shared`, and all our libraries do include cmake scripts via `../../../cmake-modules`. Therefore, as of radical renames, we are thinking of renaming `sdk/service/library/` (for the port being installed) to `_/_/_/` - https://github.com/Azure/azure-sdk-for-cpp/pull/4977. This is the most radical we can go at the moment.

**Comment by antkmsft** [comment]
@Cheney-W, if there is no further feedback, would it be possible to get this merged by the end of this week?

**Review by Cheney-W — APPROVED** [review]

**Pull request merged**